File: vol2/ch20/LeetCode2473.cpp
typedef long long LL;
struct Edge {                                     //出边类
    int vno;                                      //邻接点
    LL wt;                                       //边的权
    Edge(int v,LL w):vno(v),wt(w) {}
};
struct QNode {                  //优先队列结点类型
    int vno;                    //顶点编号
    int length;                 //路径长度
    bool operator<(const QNode& b) const {
        return length>b.length;     //length越小越优先出队
    }
};
class Solution {
    const LL INF=0x3f3f3f3f3f3f3f3f;
public:
    vector<long long> minCost(int n,vector<vector<int>>&roads,vector<int>& appleCost,int k) {
        vector<vector<Edge>> adj(n+1);      //邻接表
        for(int i=0;i<roads.size();i++) {   //遍历road建立邻接表
            int a=roads[i][0];
            int b=roads[i][1];
            LL w=(k+1)*roads[i][2];         //<a,b>:w
            adj[a].push_back(Edge(b,w));
            adj[b].push_back(Edge(a,w));
        }
        for(int i=1;i<=n;i++) {         //建超级源点的边
            adj[0].push_back(Edge(i,appleCost[i-1]));
            adj[i].push_back(Edge(0,appleCost[i-1]));
        }
        vector<LL> dist(n+1,INF);
        priority_queue<QNode> pqu;          //小根堆
        QNode e,e1;
        dist[0]=0;
        e.vno=0; e.length=0;
        pqu.push(e);
        while(!pqu.empty()) {
            int u=pqu.top().vno;pqu.pop();  //出队顶点u
            for(auto edj:adj[u]) {
                int v=edj.vno;
                LL w=edj.wt;
                if(dist[v]>dist[u]+w) {     //边松驰
                    dist[v]=dist[u]+w;
                    e1.vno=v;
                    e1.length=dist[v];
                    pqu.push(e1);
                }
            }
        }
        vector<LL> ans(n);
        for(int i=1;i<=n;i++)
            ans[i-1]=dist[i];
        return ans;
    }
};

File: vol2/ch20/LeetCode2473.py
class Solution:
    def minCost(self,n:int,roads:List[List[int]],appleCost:List[int],k:int):
        INF=0x3f3f3f3f3f3f3f3f
        adj=[[] for _ in range(0,n+1)]                    #邻接表
        for i in range(0,len(roads)):      #遍历roads建立邻接表
            a,b=roads[i][0],roads[i][1]     #<a,b>:w
            w=(k+1)*roads[i][2]
            adj[a].append([b,w])
            adj[b].append([a,w])
        for i in range(1,n+1):         #建超级源点的边
            adj[0].append([i,appleCost[i-1]])
            adj[i].append([0,appleCost[i-1]])
        dist=[INF for _ in range(0,n+1)]
        dist[0]=0
        pqu=[]                          #定义小根堆pqu,元素为[length,vno]
        heapq.heappush(pqu,[0,0])      #源点元素进队
        while pqu:
            e=heapq.heappop(pqu)        #出队元素e
            u=e[1]                      #出队顶点u
            for edj in adj[u]:
                v,w=edj[0],edj[1]
                if dist[v]>dist[u]+w:    #边松驰
                    dist[v]=dist[u]+w
                    heapq.heappush(pqu,[dist[v],v])      #顶点v元素进队
        ans=[0]*n
        for i in range(1,n+1):
            ans[i-1]=dist[i]
        return ans
